pop-api/src/error.rs

```diff
@@ -0,0 +1,405 @@
+	// Converts single nested errors that are known to the Pop API as unit errors into `Other`.
+	if UNIT_ERRORS.contains(&encoded_error[0]) && encoded_error[1..].iter().any(|x| *x != 0u8) {
+		encoded_error[..].rotate_right(1);
+		encoded_error[0] = 0u8;
+	// Converts double nested errors that are known to the Pop API as single nested errors into
+	// `Other`.
+	} else if SINGLE_NESTED_ERRORS.contains(&encoded_error[0])
+		&& encoded_error[2..].iter().any(|x| *x != 0u8)
+	{
+		encoded_error[..].rotate_right(1);
+		encoded_error[0] = 0u8;
+	} else if DOUBLE_NESTED_ERRORS.contains(&encoded_error[0])
+		&& encoded_error[3..].iter().any(|x| *x != 0u8)
+	{
+		encoded_error[..].rotate_right(1);
+		encoded_error[0] = 0u8;
+	}
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+#[repr(u8)]
+pub enum Error {
+	/// Some unknown error occurred. Go to the Pop API docs section `Pop API error`.
+	Other {
+		// Index within the `DispatchError`
+		dispatch_error_index: u8,
+		// Index within the `DispatchError` variant.
+		error_index: u8,
+		// Index for further nesting, e.g. pallet error.
+		error: u8,
+	} = 0,
+	/// Failed to lookup some data.
+	CannotLookup = 1,
+	/// A bad origin.
+	BadOrigin = 2,
+	/// A custom error in a module.
+	Module {
+		index: u8,
+		error: u8,
+	} = 3,
+	/// At least one consumer is remaining so the account cannot be destroyed.
+	ConsumerRemaining = 4,
+	/// There are no providers so the account cannot be created.
+	NoProviders = 5,
+	/// There are too many consumers so the account cannot be created.
+	TooManyConsumers = 6,
+	/// An error to do with tokens.
+	Token(TokenError) = 7,
+	/// An arithmetic error.
+	Arithmetic(ArithmeticError) = 8,
+	/// The number of transactional layers has been reached, or we are not in a transactional
+	/// layer.
+	Transactional(TransactionalError) = 9,
+	/// Resources exhausted, e.g. attempt to read/write data which is too large to manipulate.
+	Exhausted = 10,
+	/// The state is corrupt; this is generally not going to fix itself.
+	Corruption = 11,
+	/// Some resource (e.g. a preimage) is unavailable right now. This might fix itself later.
+	Unavailable = 12,
+	/// Root origin is not allowed.
+	RootNotAllowed = 13,
+	DecodingFailed = 255,
+}
+
+// Unit `Error` variants.
+// (variant: index):
+// - CannotLookup: 1,
+// - BadOrigin: 2,
+// - ConsumerRemaining: 4,
+// - NoProviders: 5,
+// - TooManyConsumers: 6,
+// - Exhausted: 10,
+// - Corruption: 11,
+// - Unavailable: 12,
+// - RootNotAllowed: 13,
+// - DecodingFailed: 255,
+const UNIT_ERRORS: [u8; 10] = [1, 2, 4, 5, 6, 10, 11, 12, 13, 255];
+
+// Single nested `Error` variants.
+// (variant: index):
+// - Token: 7,
+// - Arithmetic: 8,
+// - Transaction: 9,
+const SINGLE_NESTED_ERRORS: [u8; 3] = [7, 8, 9];
+
+// Double nested `Error` variants
+// (variant: index):
+// - Module: 3,
+const DOUBLE_NESTED_ERRORS: [u8; 1] = [3];
```

Comment:
Implementing a TryFrom<u8> would result in the same amount of code duplication (necessary in this case), however, the TryFrom<u8> will improve the code readability

Reply:
It can be even simpler. You are using a `#[repr(u8)]` which means to convert from an Enum to a u8 simply can be done like (for example): `Error::Other as u8`

Unfortunately, going the other way (u8 -> enum) requires the TryFrom<u8> and individual mappings.

Reply:
No that is not possible